File: python3-sdk/semantria/session.py
# -*- coding: utf-8 -*-

from semantria.authrequest import *
from semantria.jsonserializer import * 
from semantria.mapping import *

class Session:
    host = 'https://api21.semantria.com'
    wrapperName = 'Python_3.x'
    
    def __init__(self, consumerKey, consumerSecret, serializer=None, applicationName=None):
        self.consumerKey = consumerKey
        self.consumerSecret = consumerSecret

        if applicationName:
            self.applicationName = '{0}.{1}'.format(applicationName, self.wrapperName)
        else:
            self.applicationName = self.wrapperName
            
        if serializer:
            self.serializer = serializer
            self.format = serializer.gettype()
        else:
            #set default json serializer
            self.serializer = JsonSerializer()
            self.format = self.serializer.gettype()

        #Events
        self.Request = SessionEvent(self) 
        self.Response = SessionEvent(self) 
        self.Error = SessionEvent(self) 
        self.DocsAutoResponse = SessionEvent(self) 
        self.CollsAutoResponse = SessionEvent(self) 

    def registerSerializer(self, serializer):
        if serializer:
            self.serializer = serializer
            self.format = serializer.gettype()
        else :
            raise Error('Parameter not found: %s' % serializer)
    
    def createUpdateProxy(self):
        return {"added":[], "removed":[], "cloned":[]}

    def getStatus(self):
        url = '{0}/status.{1}'.format(self.host, self.format)
        return self._runRequest("GET", url, "get_status")

    def verifySubscription(self):
        url = '{0}/subscription.{1}'.format(self.host, self.format)
        return self._runRequest("GET", url, "get_subscription")
        
    def getConfigurations(self):
        url = '{0}/configurations.{1}'.format(self.host, self.format)
        result = self._runRequest("GET", url, "get_configurations")
        if result is None:
            result = []
        return result
    
    def updateConfigurations(self, updates):
        url = '{0}/configurations.{1}'.format(self.host, self.format)
        wrapper = self._getTypeWrapper("update_configurations")
        data = self.serializer.serialize(updates, wrapper);
        return self._runRequest("POST", url, None, data)
           
    def getBlacklist(self, configId=None):
        if configId:
            url = '{0}/blacklist.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/blacklist.{1}'.format(self.host, self.format)
        
        result = self._runRequest("GET", url, "get_blacklist")
        if result is None:
            result = []
        return result
    
    def updateBlacklist(self, updates, configId=None):
        if configId:
            url = '{0}/blacklist.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/blacklist.{1}'.format(self.host, self.format)
        
        wrapper = self._getTypeWrapper("update_blacklist")
        data = self.serializer.serialize(updates, wrapper);
        return self._runRequest("POST", url, None, data)
         
    def getCategories(self, configId=None):
        if configId:
            url = '{0}/categories.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/categories.{1}'.format(self.host, self.format)
        
        result = self._runRequest("GET", url, "get_categories")
        if result is None:
            result = []
        return result
   
    def updateCategories(self, updates, configId=None):
        if configId:
            url = '{0}/categories.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/categories.{1}'.format(self.host, self.format)
        
        wrapper = self._getTypeWrapper("update_categories")
        data = self.serializer.serialize(updates, wrapper);
        return self._runRequest("POST", url, None, data)

    def getQueries(self, configId=None):
        if configId:
            url = '{0}/queries.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/queries.{1}'.format(self.host, self.format)
        
        result = self._runRequest("GET", url, "get_queries")
        if result is None:
            result = []
        return result
    
    def updateQueries(self, updates, configId=None):
        if configId:
            url = '{0}/queries.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/queries.{1}'.format(self.host, self.format)

        wrapper = self._getTypeWrapper("update_queries")
        data = self.serializer.serialize(updates, wrapper);
        return self._runRequest("POST", url, None, data)

    def getSentimentPhrases(self, configId=None):
        if configId:
            url = '{0}/sentiment.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/sentiment.{1}'.format(self.host, self.format)
        
        result = self._runRequest("GET", url, "get_sentiment_phrases")
        if result is None:
            result = []
        return result
    
    def updateSentimentPhrases(self, updates, configId=None):
        if configId:
            url = '{0}/sentiment.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/sentiment.{1}'.format(self.host, self.format)

        wrapper = self._getTypeWrapper("update_sentiment_phrases")
        data = self.serializer.serialize(updates, wrapper);
        return self._runRequest("POST", url, None, data)

    def getEntities(self, configId=None):
        if configId:
            url = '{0}/entities.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/entities.{1}'.format(self.host, self.format)
        
        result = self._runRequest("GET", url, "get_entities")
        if result is None:
            result = []
        return result
    
    def updateEntities(self, updates, configId=None):
        if configId:
            url = '{0}/entities.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/entities.{1}'.format(self.host, self.format)

        wrapper = self._getTypeWrapper("update_entities")
        data = self.serializer.serialize(updates, wrapper);
        return self._runRequest("POST", url, None, data)
    
    def queueDocument(self, task, configId=None):
        if configId:
            url = '{0}/document.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/document.{1}'.format(self.host, self.format)

        wrapper = self._getTypeWrapper("queue_document")
        data = self.serializer.serialize(task, wrapper);
        result = self._runRequest("POST", url, "get_processed_documents", data)
        if (result != None and not isinstance(result, int)):
            self.DocsAutoResponse(result)
            return 200
        else:
            return result
    
    def queueBatchOfDocuments(self, batch, configId=None):
        if configId:
            url = '{0}/document/batch.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/document/batch.{1}'.format(self.host, self.format)

        wrapper = self._getTypeWrapper("queue_batch_documents")
        data = self.serializer.serialize(batch, wrapper);

        result = self._runRequest("POST", url, "get_processed_documents", data)
        if (result != None and not isinstance(result, int)):
            self.DocsAutoResponse(result)
            return 200
        else:
            return result

    def getDocument(self, id, configId=None):
        if id:
            if configId:
                url = '{0}/document/{1}.{2}?config_id={3}'.format(self.host, id, self.format, configId)
            else:
                url = '{0}/document/{1}.{2}'.format(self.host, id, self.format)
        else:
            raise Error('Parameter not found: %s' % id) 

        return self._runRequest("GET", url, "get_document")

    def cancelDocument(self, id, configId=None):
        if id:
            if configId:
                url = '{0}/document/{1}.{2}?config_id={3}'.format(self.host, id, self.format, configId)
            else:
                url = '{0}/document/{1}.{2}'.format(self.host, id, self.format)
        else:
            raise Error('Parameter not found: %s' % id) 

        return self._runRequest("DELETE", url)

    def getProcessedDocuments(self, configId=None):
        if configId:
            url = '{0}/document/processed.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/document/processed.{1}'.format(self.host, self.format)
        
        result = self._runRequest("GET", url, "get_processed_documents")
        if result is None:
            result = []
        return result
    
    def queueCollection(self, task, configId=None):
        if configId:
            url = '{0}/collection.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/collection.{1}'.format(self.host, self.format)

        wrapper = self._getTypeWrapper("queue_collection")
        data = self.serializer.serialize(task, wrapper);
        result = self._runRequest("POST", url, "get_processed_collections", data)
        if (result != None and not isinstance(result, int)):
            self.CollsAutoResponse(result)
            return 200
        else:
            return result

    def getCollection(self, id, configId=None):
        if id:
            if configId:
                url = '{0}/collection/{1}.{2}?config_id={3}'.format(self.host, id, self.format, configId)
            else:
                url = '{0}/collection/{1}.{2}'.format(self.host, id, self.format)
        else:
            raise Error('Parameter not found: %s' % id) 

        return self._runRequest("GET", url, "get_collection")

    def cancelCollection(self, id, configId=None):
        if id:
            if configId:
                url = '{0}/collection/{1}.{2}?config_id={3}'.format(self.host, id, self.format, configId)
            else:
                url = '{0}/collection/{1}.{2}'.format(self.host, id, self.format)
        else:
            raise Error('Parameter not found: %s' % id) 

        return self._runRequest("DELETE", url)

    def getProcessedCollections(self, configId=None):
        if configId:
            url = '{0}/collection/processed.{1}?config_id={2}'.format(self.host, self.format, configId)
        else:
            url = '{0}/collection/processed.{1}'.format(self.host, self.format)
        
        result = self._runRequest("GET", url, "get_processed_collections")
        if result is None:
            result = []
        return result
      
    def _getTypeHandler(self, type):
        if self.serializer.gettype() == "json":
            return None

        #only for xml serializer
        if type == "get_status":
            return GetStatusHandler()
        elif type == "get_subscription":
            return GetSubscriptionHandler()
        elif type == "get_configurations":
            return GetConfigurationsHandler()
        elif type == "get_blacklist":
            return GetBlacklistHandler()
        elif type == "get_categories":
            return GetCategoriesHandler()
        elif type == "get_queries":
            return GetQueriesHandler()
        elif type == "get_sentiment_phrases":
            return GetSentimentPhrasesHandler()
        elif type == "get_entities":
            return GetEntitiesHandler()
        elif type == "get_document":
            return GetDocumentHandler()
        elif type == "get_processed_documents":
            return GetProcessedDocumentsHandler()
        elif type == "get_collection":
            return GetCollectionHandler()
        elif type == "get_processed_collections":
            return GetProcessedCollectionsHandler()
        else:
            return None

    def _getTypeWrapper(self, type):
        if self.serializer.gettype() == "json":
            return None

        #only for xml serializer
        if type == "update_configurations":
            return {"root":"configurations", "added":"configuration", "removed":"configuration"}
        elif type == "update_blacklist":
            return {"root":"blacklist", "added":"item", "removed":"item"}
        elif type == "update_categories":
            return {"root":"categories", "added":"category", "removed":"category", "samples":"sample"}
        elif type == "update_queries":
            return {"root":"queries", "added":"query", "removed":"query"}
        elif type == "update_sentiment_phrases":
            return {"root":"phrases", "added":"phrase", "removed":"phrase"}
        elif type == "update_entities":
            return {"root":"entities", "added":"entity", "removed":"entity"}
        elif type == "queue_document":
            return {"root":"document"}
        elif type == "queue_batch_documents":
            return {"root":"documents", "item":"document"}
        elif type == "queue_collection":
            return {"root":"collection", "documents":"document"}
        else:
            return None

    def _runRequest(self, method, url, type=None, postData=None):
        request = AuthRequest(self.consumerKey, self.consumerSecret, self.applicationName)
        self.Request({"method":method, "url":url, "message":postData}) 
        response = request.authWebRequest(method, url, postData);
        self.Response({"status":response["status"], "reason":response["reason"], "message":response["data"]}) 
        
        status = response["status"]
        message = response["reason"]
        
        if response["data"] != "":
            message = response["data"]

        if method == "DELETE":
            if (status == 200 or status == 202):
                return status
            else:
                self._resolveError(status, message)
                return status
        else:
            if status == 200:
                handler = self._getTypeHandler(type) 
                message = self.serializer.deserialize(response["data"], handler)
                return message
            elif status == 202:
                if method == "POST":
                    return status
                else:
                    return None
            else:
                self._resolveError(status, message)
     
    def _resolveError(self, status, message=None):
        if status == 400 or status == 401 or status == 402 or status == 403 or status == 406 or status == 500:
            self.Error({"status":status, "message":message})
        else:
            raise http.client.HTTPException(status, message)

class SessionEvent: 
    def __init__(self, sender): 
        self.handlers = set()
        self.sender = sender 
 
    def handle(self, handler): 
        self.handlers.add(handler) 
        return self 
 
    def unhandle(self, handler): 
        try: 
            self.handlers.remove(handler) 
        except: 
            raise ValueError("Handler is not handling this event, so cannot unhandle it.") 
        return self 
 
    def fire(self, *args, **kargs): 
        for handler in self.handlers: 
            handler(self.sender, *args, **kargs) 
 
    def getHandlerCount(self): 
        return len(self.handlers) 
 
    __iadd__ = handle 
    __isub__ = unhandle 
    __call__ = fire 
    __len__  = getHandlerCount

File: cpp-sdk/src/Session.cpp
#include "Session.h"


Session::Session() {
    init((string)"", (string)"", (void*)NULL, "");
}

Session::Session(string consumerKey, string consumerSecret, void* serializer, string applicationName) {
    init(consumerKey, consumerSecret, serializer, applicationName);
}

void Session::init(string consumerKey, string consumerSecret, void* serializer, string applicationName) {
    this->consumerKey = consumerKey;
    this->consumerSecret = consumerSecret;
    this->callbackHandler = NULL;
    this->serializer = NULL;
    SetApplicationName(applicationName);

    if (NULL != serializer) {
        RegisterSerializer(serializer);
    }

    curl_global_init(CURL_GLOBAL_ALL);
}

void Session::SetApplicationName(string applicationName) {
    if ("" != applicationName) {
        this->applicationName = applicationName + "." + wrapperName;
    } else {
        this->applicationName = wrapperName;
    }
}

Session::~Session() {
	if (NULL != serializer) {
	    delete serializer;
	}

	if (NULL != callbackHandler) {
        delete callbackHandler;
	}

	curl_global_cleanup();
}

Session* Session::CreateSession(string consumerKey, string consumerSecret) {
    return Session::CreateSession(consumerKey, consumerSecret, new XmlSerializer());
}

Session* Session::CreateSession(string consumerKey, string consumerSecret, string applicationName) {
    return Session::CreateSession(consumerKey, consumerSecret, new XmlSerializer(), applicationName);
}

Session* Session::CreateSession(string consumerKey, string consumerSecret, void* serializer) {
    return new Session(consumerKey, consumerSecret, serializer, "");
}

Session* Session::CreateSession(string consumerKey, string consumerSecret, void* serializer, string applicationName) {
    return new Session(consumerKey, consumerSecret, serializer, applicationName);
}

void Session::SetCallbackHandler(CallbackHandler* callbackHandler) {
    this->callbackHandler = callbackHandler;
}

void Session::RegisterSerializer(void* serializer) {
    this->serializer = serializer;
    // We can use casting to any Serializer derived class, "type" should be constant
    format = (static_cast<XmlSerializer*>(serializer))->GetType();
}

// End-point calls
ServiceStatus* Session::GetStatus() {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, "", applicationName);
    QueryMethod method = GET;
    string url = host + "/status." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);

    ServiceStatus* result = new ServiceStatus();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        }
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

Subscription* Session::VerifySubscription() {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, "", applicationName);
    QueryMethod method = GET;
    string url = host + "/subscription." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);

    Subscription* result = new Subscription();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        }
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

vector<string>* Session::GetBlacklist() {
    return GetBlacklist("");
}

vector<string>* Session::GetBlacklist(string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/blacklist." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);

    vector<string>* result = new vector<string>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_Blacklist* stub = new Stub_Blacklist();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetBlacklistItems();

        // @TODO don't forget that items are stored in stub, so if
        // we delete items in stub's destructor - all items from result
        // will be removed too.
        // maybe we should use smart pointer?
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

int Session::UpdateBlacklist(UpdateProxy<Blacklisted>* proxy) {
    return UpdateBlacklist(proxy, "");
}

int Session::UpdateBlacklist(UpdateProxy<Blacklisted>* proxy, string configId) {
    return UpdateItems("blacklist", proxy, configId);
}

template <class T>
int Session::UpdateItems(string path, UpdateProxy<T>* proxy, string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = POST;
    string url = host + "/" + path + "." + format;
    cout << endl << "url: " << url << endl;
    string body = "";
    if (format == "xml") {
        body = (static_cast<XmlSerializer*>(serializer))->Serialize(proxy);
    } else {
        body = (static_cast<JsonSerializer*>(serializer))->Serialize(proxy);
    }

    int result = authRequest->authWebRequest(url, method, body);

    HandleRequest(POST, url, body);
    HandleConfigurationResponse(result, authRequest);

    delete authRequest;
    return result;
}

vector<Category*>* Session::GetCategories() {
    return GetCategories("");
}

vector<Category*>* Session::GetCategories(string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/categories." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);

    vector<Category*>* result = new vector<Category*>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_Categories* stub = new Stub_Categories();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetCategories();
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

int Session::UpdateCategories(UpdateProxy<Category>* proxy) {
    return UpdateCategories(proxy, "");
}

int Session::UpdateCategories(UpdateProxy<Category>* proxy, string configId) {
    return UpdateItems("categories", proxy, configId);
}

vector<Configuration*>* Session::GetConfigurations() {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, "", applicationName);
    QueryMethod method = GET;
    string url = host + "/configurations." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    vector<Configuration*>* result = new vector<Configuration*>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_Configurations* stub = new Stub_Configurations();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetConfigurations();
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

int Session::UpdateConfigurations(UpdateProxy<Configuration>* proxy) {
    return UpdateConfigurations(proxy, "");
}

int Session::UpdateConfigurations(UpdateProxy<Configuration>* proxy, string configId) {
    return UpdateItems("configurations", proxy, configId);
}

vector<UserEntity*>* Session::GetEntities() {
    return GetEntities("");
}

vector<UserEntity*>* Session::GetEntities(string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/entities." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    vector<UserEntity*>* result = new vector<UserEntity*>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_Entities* stub = new Stub_Entities();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetEntities();
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

int Session::UpdateEntities(UpdateProxy<UserEntity>* proxy) {
    return UpdateEntities(proxy, "");
}

int Session::UpdateEntities(UpdateProxy<UserEntity>* proxy, string configId) {
    return UpdateItems("entities", proxy, configId);
}

vector<Query*>* Session::GetQueries() {
    return GetQueries("");
}

vector<Query*>* Session::GetQueries(string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/queries." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    vector<Query*>* result = new vector<Query*>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_Queries* stub = new Stub_Queries();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetQueries();
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

int Session::UpdateQueries(UpdateProxy<Query>* proxy) {
    return UpdateQueries(proxy, "");
}

int Session::UpdateQueries(UpdateProxy<Query>* proxy, string configId) {
    return UpdateItems("queries", proxy, configId);
}

vector<SentimentPhrase*>* Session::GetSentimentPhrases() {
    return GetSentimentPhrases("");
}

vector<SentimentPhrase*>* Session::GetSentimentPhrases(string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/sentiment." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    vector<SentimentPhrase*>* result = new vector<SentimentPhrase*>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_SentimentPhrase* stub = new Stub_SentimentPhrase();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetPhrases();
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

int Session::UpdateSentimentPhrases(UpdateProxy<SentimentPhrase>* proxy) {
    return UpdateSentimentPhrases(proxy, "");
}

int Session::UpdateSentimentPhrases(UpdateProxy<SentimentPhrase>* proxy, string configId) {
    return UpdateItems("sentiment", proxy, configId);
}

DocAnalyticData* Session::GetDocument(string documentId) {
    return GetDocument(documentId, "");
}

DocAnalyticData* Session::GetDocument(string documentId, string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/document/" + documentId + "." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    DocAnalyticData* result = new DocAnalyticData();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        }
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

CollAnalyticData* Session::GetCollection(string collectionId) {
    return GetCollection(collectionId, "");
}

CollAnalyticData* Session::GetCollection(string collectionId, string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/collection/" + collectionId + "." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    CollAnalyticData* result = new CollAnalyticData();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), result);
        }
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

vector<DocAnalyticData*>* Session::GetProcessedDocuments() {
    return GetProcessedDocuments("");
}

vector<DocAnalyticData*>* Session::GetProcessedDocuments(string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/document/processed." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    vector<DocAnalyticData*>* result = new vector<DocAnalyticData*>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_DocAnalyticDatas* stub = new Stub_DocAnalyticDatas();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetAnalyticData();
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

vector<CollAnalyticData*>* Session::GetProcessedCollections() {
    return GetProcessedCollections("");
}

vector<CollAnalyticData*>* Session::GetProcessedCollections(string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = GET;
    string url = host + "/collection/processed." + format;
    cout << endl << "url: " << url << endl;
    int status = authRequest->authWebRequest(url, method);
    vector<CollAnalyticData*>* result = new vector<CollAnalyticData*>();
    if (status <= 202 && "" != authRequest->GetReceivedData()) {
        Stub_CollAnalyticDatas* stub = new Stub_CollAnalyticDatas();
        if (format == "xml") {
            (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        } else {
            (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
        }
        result = stub->GetAnalyticData();
        delete stub;
    }
    HandleRequest(method, url, "");
    HandleConfigurationResponse(status, authRequest);

    delete authRequest;

    return result;
}

int Session::QueueDocument(Document* document) {
    return QueueDocument(document, "");
}

int Session::QueueDocument(Document* document, string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = POST;
    string url = host + "/document." + format;
    cout << endl << "url: " << url << endl;

    string body = "";
    if (format == "xml") {
        body = (static_cast<XmlSerializer*>(serializer))->Serialize(document);
    } else {
        body = (static_cast<JsonSerializer*>(serializer))->Serialize(document);
    }
    int result = authRequest->authWebRequest(url, method, body);

    HandleDocAutoResponse(result, authRequest);
    HandleRequest(method, url, body);

    delete authRequest;
    return result;
}

int Session::QueueBatch(vector<Document*>* documents) {
    return QueueBatch(documents, "");
}

int Session::QueueBatch(vector<Document*>* documents, string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = POST;
    string url = host + "/document/batch." + format;
    cout << endl << "url: " << url << endl;
    Stub_Tasks* documentsStub = new Stub_Tasks(documents);
    string body = "";
    if (format == "xml") {
        body = (static_cast<XmlSerializer*>(serializer))->Serialize(documentsStub);
    } else {
        body = (static_cast<JsonSerializer*>(serializer))->Serialize(documentsStub);
    }
    delete documentsStub;

    int result = authRequest->authWebRequest(url, method, body);

    HandleDocAutoResponse(result, authRequest);
    HandleRequest(method, url, body);

    delete authRequest;
    return result;
}

int Session::QueueCollection(Collection* collection) {
    return QueueCollection(collection, "");
}

int Session::QueueCollection(Collection* collection, string configId) {
    AuthRequest* authRequest = new AuthRequest(consumerKey, consumerSecret, configId, applicationName);
    QueryMethod method = POST;
    string url = host + "/collection." + format;
    cout << endl << "url: " << url << endl;
    string body = "";
    if (format == "xml") {
        body = (static_cast<XmlSerializer*>(serializer))->Serialize(collection);
    } else {
        body = (static_cast<JsonSerializer*>(serializer))->Serialize(collection);
    }
    int result = authRequest->authWebRequest(url, method, body);

    HandleCollAutoResponse(result, authRequest);
    HandleRequest(method, url, body);

    delete authRequest;

    return result;
}

void Session::HandleConfigurationResponse(int status, AuthRequest* authRequest) {
    if (NULL != callbackHandler) {
        if (status <= 202) {
            callbackHandler->OnResponse(this, new ResponseArgs(status, authRequest->GetReceivedData()));
        } else {
            callbackHandler->OnError(this, new ResponseArgs(status, authRequest->GetReceivedData()));
        }
    }
}

void Session::HandleDocAutoResponse(int status, AuthRequest* authRequest) {
    if (NULL != callbackHandler) {
        HandleConfigurationResponse(status, authRequest);

        if (status <= 202 && "" != authRequest->GetReceivedData()) {
            Stub_DocAnalyticDatas* stub = new Stub_DocAnalyticDatas();
            if (format == "xml") {
                (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
            } else {
                (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
            }
            callbackHandler->OnDocsAutoResponse(this, stub->GetAnalyticData());
            delete stub;
        }
    }
}

void Session::HandleCollAutoResponse(int status, AuthRequest* authRequest) {
    if (NULL != callbackHandler) {
        HandleConfigurationResponse(status, authRequest);

        if (status <= 202 && "" != authRequest->GetReceivedData()) {
            Stub_CollAnalyticDatas* stub = new Stub_CollAnalyticDatas();
            if (format == "xml") {
                (static_cast<XmlSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
            } else {
                (static_cast<JsonSerializer*>(serializer))->Deserialize(authRequest->GetReceivedData(), stub);
            }
            callbackHandler->OnCollsAutoResponse(this, stub->GetAnalyticData());
            delete stub;
        }
    }
}

void Session::HandleRequest(QueryMethod method, string url, string message) {
    if (NULL != callbackHandler) {
        callbackHandler->OnRequest(this, new RequestArgs(EnumsHelper::GetQueryMethodAsString(method), url, message));
    }
}

const string Session::host = "https://api21.semantria.com";
const string Session::wrapperName = "Cpp";
